feat: changed find users method to use new search method

toymachiner62 opened the pull request:
Initial pass at structured style of finding users using the new search users method.

### TODO:
- [x] update docs https://developers.catalytic.com/docs/search-users-java
- [x] update api to return `isTeamAdmin`, `isDeactivated`, `isLockedOut`, and `createdDate` for User's https://github.com/catalyticlabs/CatalyticSDKAPI/issues/237 (@benblair)
- [x] Remove unnecessary methods from `GuidSearchExpression` https://github.com/catalyticlabs/CatalyticSDKAPI/issues/232 (@benblair)

toymachiner62 commented:
Doing a search using `and` or `or` seems extremely complicated and cumbersome using this approach.

Here's an integration test that searches for `id = f62d0251-cb1b-4ad2-aa51-deb8461be7db` OR `email = tcaflisch@catalytic.com`

```java
@Test
    public void itShouldFindUserByIdOrEmail() throws Exception {
        CatalyticClient catalytic = new CatalyticClient("");

        List<UserSearchClause> or = new ArrayList<>();
        UserSearchClause idClause = new UserSearchClause();
        UserSearchClause emailClause = new UserSearchClause();
        GuidSearchExpression id = new GuidSearchExpression();
        id.setIsEqualTo(UUID.fromString("f62d0251-cb1b-4ad2-aa51-deb8461be7db"));
        StringSearchExpression email = new StringSearchExpression();
        email.setIsEqualTo("tcaflisch@catalytic.com");
        idClause.setId(id);
        emailClause.setEmail(email);
        or.add(idClause);
        or.add(emailClause);

        UsersPage results = catalytic.users().find(null, or, null, null, null, null, null, null, null, null, null);
        assertThat(results.getUsers()).isNotEmpty();
        assertThat(results.getNextPageToken()).isNotNull();
        assertThat(results.getCount()).isGreaterThan(0);

        for (User user : results.getUsers()) {
            System.out.println(user.toString());
        }
    }
```

nrod80 commented:
> ```java
> List<UserSearchClause> or = new ArrayList<>();
> 
> UserSearchClause idClause = new UserSearchClause();
> GuidSearchExpression id = new GuidSearchExpression();
> id.setIsEqualTo(UUID.fromString("f62d0251-cb1b-4ad2-aa51-deb8461be7db"));
> idClause.setId(id);
> or.add(idClause);
> 
> UserSearchClause emailClause = new UserSearchClause();
> StringSearchExpression email = new StringSearchExpression();
> email.setIsEqualTo("tcaflisch@catalytic.com");
> emailClause.setEmail(email);
> or.add(emailClause);
> ```

I might be in the minority but I find directly writing JSON to be much more intuitive than using built-in SDK types to generate filter expressions. For instance, the above seems like a lot of code to produce `{ or: [{ id: "f62d0251-cb1b-4ad2-aa51-deb8461be7db" }, { email: "tcaflisch@catalytic.com" }] }`.

Would it be possible to add an overload to support something like the following syntax?
```java
string filter = "{ \"or\": [{ \"id\": \"f62d0251-cb1b-4ad2-aa51-deb8461be7db\" }, { \"email\": \"tcaflisch@catalytic.com\" }] }";
catalytic.users().find(filter);
```

I'm not very familiar with JSON parsing in Java, but we could possibly perform the parsing of a JSON filter expression server-side (which would also ensure consistency across SDKs and reduce duplicative implementations).

benblair commented:
For the C# version, I am working on something like this

```
UserSearchClause.Or(
    UserSearchClause.Id("f62d0251-cb1b-4ad2-aa51-deb8461be7db"),
    UserSearchClause.Email("tcaflisch@catalytic.com")
);
```

In C# I have a few more options for syntactic sugar, but the above should work in Java with static factory methods for each of the clause fields.

toymachiner62 commented:
> > ```java
> > List<UserSearchClause> or = new ArrayList<>();
> > 
> > UserSearchClause idClause = new UserSearchClause();
> > GuidSearchExpression id = new GuidSearchExpression();
> > id.setIsEqualTo(UUID.fromString("f62d0251-cb1b-4ad2-aa51-deb8461be7db"));
> > idClause.setId(id);
> > or.add(idClause);
> > 
> > UserSearchClause emailClause = new UserSearchClause();
> > StringSearchExpression email = new StringSearchExpression();
> > email.setIsEqualTo("tcaflisch@catalytic.com");
> > emailClause.setEmail(email);
> > or.add(emailClause);
> > ```
> 
> I might be in the minority but I find directly writing JSON to be much more intuitive than using built-in SDK types to generate filter expressions. For instance, the above seems like a lot of code to produce `{ or: [{ id: "f62d0251-cb1b-4ad2-aa51-deb8461be7db" }, { email: "tcaflisch@catalytic.com" }] }`.
> 
> Would it be possible to add an overload to support something like the following syntax?
> 
> ```java
> string filter = "{ \"or\": [{ \"id\": \"f62d0251-cb1b-4ad2-aa51-deb8461be7db\" }, { \"email\": \"tcaflisch@catalytic.com\" }] }";
> catalytic.users().find(filter);
> ```
> 
> I'm not very familiar with JSON parsing in Java, but we could possibly perform the parsing of a JSON filter expression server-side (which would also ensure consistency across SDKs and reduce duplicative implementations).

I don't think this is a good approach. Mixing in JSON in a hard typed language goes against the grain quite a bit.

toymachiner62 commented:
> For the C# version, I am working on something like this
> 
> ```
> UserSearchClause.Or(
>     UserSearchClause.Id("f62d0251-cb1b-4ad2-aa51-deb8461be7db"),
>     UserSearchClause.Email("tcaflisch@catalytic.com")
> );
> ```
> 
> In C# I have a few more options for syntactic sugar, but the above should work in Java with static factory methods for each of the clause fields.

I can certainly add in some syntactic sugar like this. This looks like a much simpler approach. We just have to make sure we can still handle the complex cases (contains, betweens, etc).

toymachiner62 commented on src/main/java/org/catalytic/sdk/clients/Users.java:
```diff
@@ -75,83 +75,108 @@ public User get(String identifier) throws InternalErrorException, UserNotFoundEx
-     * @throws UnauthorizedException        If unauthorized
-     */
-    public UsersPage find(Filter filter, String pageToken, Integer pageSize) throws InternalErrorException, UnauthorizedException, AccessTokenNotFoundException {
+    public UsersPage find(
+            List<UserSearchClause> and,
+            List<UserSearchClause> or,
+            GuidSearchExpression id,
+            StringSearchExpression email,
+            StringSearchExpression fullName,
+            BooleanSearchExpression isTeamAdmin,
+            BooleanSearchExpression isDeactivated,
+            BooleanSearchExpression isLockedOut,
+            DateTimeSearchExpression createdDate,
+            String pageToken,
+            Integer pageSize) throws AccessTokenNotFoundException, InternalErrorException, UnauthorizedException {
```
I'm thinking we should scrap this overloaded method for this one.

```java
public UsersPage find(UserSearchClause userSearchClause, String pageToken, Integer pageSize)
```

toymachiner62 commented:
I simplified the searching logic and added static wrappers similarly to how you did in c# @benblair.

Here's a working integration test using an OR clause

```java
    @Test
    public void itShouldFindUserByIdOrEmail() throws Exception {
        CatalyticClient catalytic = new CatalyticClient("YmI1N2I2MGMtYWMzMS00YWEzLWIwMTctMmE3OWJhMzFiNDUxOmZmODE1ZTg2YjkxMDllNGQzODUyOGJjNWJjYTdkMzE0NDdjZWNkOGMwMmIxMTEzNmVjNWZiZTU3YzdjOWRiNDQ6Y2F0YWx5dGljLnB1c2hib3QuY29tOnYx");

        UserSearchClause searchCriteria = UsersWhere.or(
                UsersWhere.id("f62d0251-cb1b-4ad2-aa51-deb8461be7db"),
                UsersWhere.email("tcaflisch@catalytic.com")
        );

        UsersPage results = catalytic.users().find(searchCriteria);

        assertThat(results.getUsers()).isNotEmpty();
        assertThat(results.getNextPageToken()).isNotNull();
        assertThat(results.getCount()).isGreaterThan(0);

        for (User user : results.getUsers()) {
            System.out.println(user.toString());
        }
    }
```

benblair commented:
```
        UserSearchClause searchCriteria = UsersWhere.or(
                UsersWhere.id("f62d0251-cb1b-4ad2-aa51-deb8461be7db"),
                UsersWhere.email("tcaflisch@catalytic.com")
        );
```

This looks great to me

benblair commented on src/main/java/org/catalytic/sdk/clients/Users.java:
```diff
@@ -168,6 +175,107 @@ public UsersPage find(Filter filter, String pageToken, Integer pageSize) throws
+     * @throws InternalErrorException       If any error finding users
+     * @throws UnauthorizedException        If unauthorized
+     */
+    public UsersPage search() throws InternalErrorException, AccessTokenNotFoundException, UnauthorizedException {
```
In the C# client, I've added an option for doing client-side paging and fetching all pages, plus an alias for `search(GetAllPages)` called `list()`. We don't necessarily have to do the same thing for the Java client but I think it makes some use-cases a little simpler.

toymachiner62 replied:
What do you mean by `client-side paging`?

Also did you add a `list()` method to the `runs` and `tasks` search as well? Those will definitely cause performance issues.

benblair replied:
The `list()` method does the `while(nextPageToken) { fetchNextPage() }` loop in the client. So to be clear, it's not doing the paging server-side and trying to send a giant payload. 

I haven't added `list()` to `instances` or `instanceSteps` yet, but I was planning to. Granted that request could take a very long time to complete, but would that cause a performance problem server-side?

toymachiner62 replied:
I guess it wouldn't cause a slow query issue, but it would definitely make a flood of requests, which is probably ok.

benblair replied:
We do some pretty aggressive backoff in response to 429's so I _think_ we'll be ok there.

toymachiner62 replied:
Do you allow using `list()` with a query or is it simply a "fetch all users" method which simply calls search and paginates behind the scenes?

benblair replied:
Yea `list()` internally just calls search without any args and paginates to the end.

toymachiner62 replied:
Ok i added a `Users.list()` method.

toymachiner62 commented on src/main/java/org/catalytic/sdk/search/UsersWhere.java:
```diff
@@ -0,0 +1,193 @@
+     * @param createdDate   The created date to search for Users by
+     * @return              The search clause
+     */
+    public static UserSearchClause createdDate(OffsetDateTime createdDate) {
```
So you allow them to manually create a string iso date such as `2020-01-01T00:00:00.000Z`?

benblair replied:
Exactly

toymachiner62 replied:
Added
